### PR TITLE
Add TIR guards.

### PR DESCRIPTION
**Pull request by vext01**
For now we implement only enough guards to cover the behaviours of our SIR terminators.

I've had to make a few changes to our terminators in light of some subtle MIR semantics which I initially missed.

I don't like the name `OtherInteger` as a guard variant name. Can you think of a better one? It should capture that we are guarding that the value of a local is not a member of a list of integer values.

Companion PR: https://github.com/softdevteam/ykrustc/pull/33

### CODE REVIEW

**Comment by vext01** [comment]
I should explain, that this is a necessary step for trace interpretation in TIRI. I'm hoping to get this working outside of ykrt for now (limiting scope and all that ;) )

File: yktrace/src/tir.rs [hunk 1]
```diff
@@ -87,7 +87,39 @@ impl TirTrace {
...
+                | Terminator::Call { .. }
+                | Terminator::Unimplemented(_) => None,
+                Terminator::Unreachable => panic!("Traced unreachable code"),
+                Terminator::SwitchInt {
+                    local,
+                    ref values,
+                    ref target_bbs,
+                    otherwise_bb
+                } => {
+                    // Peek at the next block in the trace to see which outgoing edge was taken and
```
Comment (ltratt) [inline yktrace/src/tir.rs]: Do we know that there's always another block in the trace?

Comment (vext01) [reply yktrace/src/tir.rs]: I'm working on the assumption that there is. I don't think we can stop tracing on a `SwitchInt`. Either way, it'll crash if there isn't one, so at least it's not a silent failure.

What would you prefer?

Comment (ltratt) [reply yktrace/src/tir.rs]: If that's a valid assumption, let's document it. If we're unsure, I suggest an `assert` which will at least make clear to someone that we're not entirely sure if this code is correct or not.

Comment (vext01) [reply yktrace/src/tir.rs]: I agree. Just testing the change now.

**Comment by vext01** [comment]
Try that last push.

**Comment by ltratt** [comment]
Please squash.

**Comment by vext01** [comment]
splat

**Comment by vext01** [comment]
```
---- tir::tests::nonempty_tir_trace stdout ----
thread 'tir::tests::nonempty_tir_trace' panicked at 'called `Result::unwrap()` on an `Err` value: Syntax("invalid length 3, expected struct variant Terminator::SwitchInt with 4 elements")', src/libcore/result.rs:999:5
```

The ykrustc tarball we use for testing used an older ykpack.

We have a circular dependency. We can't update ykrustc without updating yk...

Any ideas on what to do? The only think i can think of is to update ykrustc overriding the yk dependency to the feature branch to break the cycle.

**Comment by ltratt** [comment]
I didn't realise that there's a circular dependency. Should we break it once and for all? Or is that impossible?

**Comment by vext01** [comment]
Let me try to explain better. I want you to understand because I'm not sure what to do.

The problem is subtle. Ideally we want to test ykrustc and yk as one unit. That's not easy due to repo practicalities. Our workaround is to use the ykrustc tarball for the last succeeded CI build. In this case the last successdful CI build used an older ykpack with an incompatible SIR format. So boom.

**Comment by ltratt** [comment]
Can't we just delete the last CI build and this PR will therefore build everything from scratch?

**Comment by vext01** [comment]
Sadly not. Then it will crash because there is no tarball ;)

This snippet may help to explain:
https://github.com/softdevteam/yk/blob/cdc99012998c469fb665f334c051b87b6917d92a/.buildbot.sh#L5-L7

There are a couple of options I can think of:

 * We make the compiler temporarily depend on a feature branch of yk to break the cycle, then do the yk PR, and then finally adjust back the dependency in the compiler repo. This is the safest option, but involves 3 steps.

 * We manually make a tarball of the new compiler and copy it in place on the CI server (it's only the tests that make a cycle). This is easier in terms of PRs, but is racy, as another unrelated PR may overwrite our tarball at any time.

**Comment by ltratt** [comment]
> Then it will crash because there is no tarball ;)

Ah. My original intention was that it should say "if there's no tarball, rebuild everything" which would, I believe, cure this problem entirely. One reason I suggested that is precisely for cases where we get some weirdness that can be solved by rebuilding everything from scratch. Shall we do that? [Assuming that it will actually fix things! Maybe even rebuilding from source isn't enough, at which point we really will have to think again!]

**Comment by vext01** [comment]
> Ah. My original intention was that it should say "if there's no tarball, rebuild everything" which would, I believe, cure this problem entirely.

It would only work if you had some way of communicating to the yk build that it should use the as-yet unmerged ykrustc branch. Remember, we need a compiler that was built against the very yk branch that we are PRing at any given time.

**Comment by ltratt** [comment]
> It would only work if you had some way of communicating to the yk build that it should use the as-yet unmerged ykrustc branch.

Good point! Well, the easiest way to solve this in this case is to manually replace the tarball with one that's "good" and I suggest we do that. In the long run, we can perhaps think of something more clever.

**Comment by vext01** [comment]
Ack. So that's option 2 in this comment:
https://github.com/softdevteam/yk/pull/19#issuecomment-508480082

I might as well do the stage 2 test manually while i'm here then ;)

**Comment by ltratt** [comment]
OK.

**Comment by vext01** [comment]
bors try

**Comment by vext01** [comment]
You can r+ this one now.

**Pull request merged**